core/Symbols.cc

```diff
@@ -1112,6 +1112,65 @@ TypePtr Symbol::sealedSubclassesToUnion(const GlobalState &gs) const {
     return result;
 }
 
+// Record a required ancestor for this class of module
+void Symbol::recordRequiredAncestor(MutableContext ctx, SymbolRef ancestor, Loc loc) {
+    ENFORCE(this->isClassOrModule(), "Symbol is not a class or module: {}", this->show(ctx));
+
+    // We store the required ancestors into a fake property called `<required-ancestors>`
+    auto ancestors = this->findMember(ctx, Names::requiredAncestors());
+    if (!ancestors.exists()) {
+        ancestors = ctx.state.enterMethodSymbol(loc, this->ref(ctx), Names::requiredAncestors());
+        ancestors.data(ctx)->locs_.clear(); // Remove the original location
+    }
+
+    if (ancestors.data(ctx)->resultType == nullptr) {
+        vector<TypePtr> targs;
+        ancestors.data(ctx)->resultType = TupleType::build(ctx, move(targs));
+    } else {
+        // Do not require the same ancestor twice
+        for (auto elem : (cast_type<TupleType>(ancestors.data(ctx)->resultType))->elems) {
+            ENFORCE(isa_type<ClassType>(elem), "Something in requiredAncestors that's not a ClassType");
+            if (cast_type_nonnull<ClassType>(elem).symbol == ancestor) {
+                return;
+            }
+        }
```

Comment:
It might be worth trying to rephrase this in terms of `absl::any_of, just to clarify the intent.

Reply:
I changed it, and I'm not entirely sure it's clearer:

![image](https://user-images.githubusercontent.com/583144/102410778-0891ac00-3fbf-11eb-9122-cb14a33564ea.png)

For me, the comment just above is the more helpful, I let you judge.